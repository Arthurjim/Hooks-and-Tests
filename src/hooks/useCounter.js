import {useState} from 'react'

const useCounter = (intialState = 1) => {
    const [counter, setCounter] = useState(intialState);

    const increment = (factor = 1) => {
        setCounter(counter + factor )
    };

    const decrement = (factor =1) => {
        setCounter(counter - factor )
    }
    const reset = () => {
        setCounter(intialState)
    }
    return {
        counter,
        increment,
        decrement,
        reset
    }
}

export default useCounter
