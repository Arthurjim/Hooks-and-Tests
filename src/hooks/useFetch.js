import {useEffect, useRef, useState} from 'react'

const useFetch = (url) => {
    const isMounted = useRef(true);
    const [state, setState] = useState({data:null,loading:true, error:null})
    
    useEffect(()=>{

        return()=>{
            isMounted.current = false;
        }
    },[])
    useEffect(() => {

        setState({data:null,loading:true, error:null})
        
        fetch(url)
        .then(resp => resp.json())
        .then(data=>{
            setTimeout(()=>{
                if(isMounted.current){
                    setState({data:data, loading:false, error:null})
                }

            },4000)
        })
    },[url])

    return state;
}

export default useFetch
