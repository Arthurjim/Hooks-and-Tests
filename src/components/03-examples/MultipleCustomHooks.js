import React from "react";
import useCounter from "../../hooks/useCounter";
import useFetch from "../../hooks/useFetch";
import "./exampe..css";
const MultipleCustomHooks = () => {
    const {counter, increment}=useCounter(1);
   const state = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
   const { loading, data } = state;
   const { author, quote } = !!data && data[0];
   return (
      <div>
         <h1>Breaking Bad Quotes</h1>
         <hr />
         {loading ? (
            <div className="alert alert-info text-center">
               <p>Loading...</p>
            </div>
         ) : (
             
            <blockquote className="blockquote text-end">
               <p className="mb-2">{quote}</p>
               <footer className="blockquote-footer">{author}</footer>
            </blockquote>
         )}

         <button className="btn btn-primary" onClick={()=>{increment(1)}}>
             Next Quote
         </button>
      </div>
   );
};

export default MultipleCustomHooks;
