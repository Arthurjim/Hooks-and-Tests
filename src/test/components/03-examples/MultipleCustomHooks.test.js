import { shallow } from "enzyme"
import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks"
import useCounter from "../../../hooks/useCounter";
import useFetch from "../../../hooks/useFetch";
jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')

describe('Prueba al componentente MultipleCustomHooks', () => {
    beforeEach( () => {
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        })
    })
 
    test('Debe de mostrarse correctamente', () => {
        useFetch.mockReturnValue({
            data:null,
            loading:true,
            error:null
        })
        const wrapper = shallow(<MultipleCustomHooks />)
        expect(wrapper).toMatchSnapshot()
    })
    

    test('Debde de mostar la infomraciÃ³n ', () => {
        useFetch.mockReturnValue({
            data:[{
                author:'Arturo',
                quote:'TESTIIIING'
            }],
            loading:false,
            error:null
        });
        const wrapper = shallow(<MultipleCustomHooks />)
        expect(wrapper.find('.alert').exists()).toBe(false)
        expect(wrapper.find('.mb-2').text().trim()).toBe('TESTIIIING')
        expect(wrapper.find('footer').text().trim()).toBe('Arturo')


        // console.log(.wrapper.html())
    })
    
})
