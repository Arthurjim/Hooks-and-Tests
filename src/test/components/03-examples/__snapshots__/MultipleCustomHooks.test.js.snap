// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba al componentente MultipleCustomHooks Debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    <p>
      Loading...
    </p>
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote
  </button>
</div>
`;
